tests/unit_tests/service/test_authentication.py

```diff
@@ -0,0 +1,171 @@
+    )
+
+
+@pytest.fixture
+@mock.patch("requests.get")
+def mock_server_config(mock_requests_get) -> OAuthServerConfig:
+    mock_requests_get.return_value.status_code = 200
+    mock_requests_get.return_value.json.return_value = {
+        "device_authorization_endpoint": "https://example.com/device_authorization",
+        "authorization_endpoint": "https://example.com/authorization",
+        "token_endpoint": "https://example.com/token",
+        "issuer": "https://example.com",
+        "jwks_uri": "https://example.com/.well-known/jwks.json",
+        "end_session_endpoint": "https://example.com/logout",
+    }
+    return OAuthServerConfig(
+        oidc_config_url="https://auth.example.com/realms/sample/.well-known/openid-configuration",
+    )
+
+
+@pytest.fixture
+def mock_session_manager(
+    mock_client_config: OAuthClientConfig, mock_server_config: OAuthServerConfig
+) -> SessionManager | None:
+    return SessionManager.from_config(mock_server_config, mock_client_config)
+
+
+@pytest.fixture
+def mock_connected_client_config(mock_client_config: OAuthClientConfig):
+    assert isinstance(mock_client_config, CLIClientConfig)
+    with open(mock_client_config.token_file_path, "w") as token_file:
+        # base64 encoded token
+        token_file.write(
+            base64.b64encode(
+                b'{"access_token":"token","refresh_token":"refresh_token"}'
+            ).decode("utf-8")
+        )
+    return mock_client_config
+
+
+@pytest.fixture
+def mock_authenticator(mock_server_config, mock_client_config) -> Authenticator:
+    return Authenticator(mock_server_config, mock_client_config)
+
+
+@mock.patch("jwt.decode")
+@mock.patch("jwt.PyJWKClient.get_signing_key_from_jwt")
+def test_verify_token_valid(
+    mock_get_signing_key, mock_decode, mock_authenticator: Authenticator
+):
+    decode_retun_value = {"token": "valid_token", "name": "John Doe"}
+    mock_decode.return_value = decode_retun_value
+    valid_token = mock_authenticator.verify_token(decode_retun_value["token"])
+    assert valid_token
+
+
+@mock.patch("jwt.decode")
+@mock.patch("jwt.PyJWKClient.get_signing_key_from_jwt")
+def test_verify_token_invalid(
+    mock_get_signing_key, mock_decode, mock_authenticator: Authenticator
+):
+    mock_decode.side_effect = jwt.ExpiredSignatureError
+    token = "invalid_token"
+    with pytest.raises(PyJWTError):
+        mock_authenticator.verify_token(token)
+
+
+@mock.patch("jwt.decode")
+@mock.patch("jwt.PyJWKClient.get_signing_key_from_jwt")
+def test_user_info(
+    mock_get_signing_key,
+    mock_decode,
+    mock_authenticator: Authenticator,
+):
+    mock_decode.return_value = {
+        "name": "John Doe",
+        "fedid": "12345",
+    }
+    mock_authenticator.print_user_info("valid_token")
+
+
+def test_logout(
+    mock_session_manager: SessionManager, mock_connected_client_config: CLIClientConfig
+):
+    assert os.path.exists(mock_connected_client_config.token_file_path)  # type: ignore
```

Comment:
Is this ignore because `exists` expects a str? Get the string, don't be this lazy